doc/pattern.md

```diff
@@ -0,0 +1,183 @@
+---
+title: Spoon Patterns
+---
+
+Spoon patterns enables you to find code elements. A Spoon pattern is based on a one or several AST nodes, which represent the code to match, where some parts of the AST are pattern parameters. When a pattern is matched, one can access to the code matched in each pattern parameter.
+
+The main classes of Spoon patterns are those in package `spoon.pattern`:
+
+* classes: PatternBuilder, Pattern, Match, PatternBuilderHelper, PatternParameterConfigurator, InlinedStatementConfigurator 
+* eums: ConflictResolutionMode, Quantifier
+
+Example usage:
+
+```java
+Factory spoonFactory = ...
+//build a Spoon pattern
+Pattern pattern = ... build a spoon pattern. For example for an method ...
+
+//search for all occurences of the method in the root package
+pattern.forEachMatch(spoonFactory.getRootPackage(), (Match match) -> {
+	Map<String, Object> parameters = match.getParametersAsMap();
+	CtMethod<?> matchingMethod = match.getMatchingElement(CtMethod.class);
+	String aNameOfMatchedMethod = parameters.get("methodName");
+	...
+});
+```
+
+## PatternBuilder
+
+To create a Spoon pattern, one must use `PatternBuilder`, which takes AST nodes as input, and **pattern parameters** are defined.
+
+
+```java
+// creates pattern from the body of method "matcher1"
+Pattern t = PatternBuilder.create(
+    new PatternBuilderHelper(fooClass).setBodyOfMethod("matcher1").getPatternElements())
+    .configurePatternParameters()
+    .build();
+```
+
+
+This pattern matches all statements of the body of method `statement`, ie. a precondition to check that a list  is smaller than a certain size. 
+This pattern has  one single pattern parameter called `_col_`, which is automatically considered as a pattern parameter because it is declared outside of the AST node. This automatic configuration happens when `configurePatternParameters` is called.
```

Comment:
The keywords `statement` and `_col_` are wrong with updated example

Reply:
could you fix them directly? that would likely be more effective.

Reply:
I started to do it before I wrote this comment, but then I had feeling like there is too many changes and I do not know your vision, so my fix might ho in wrong direction:
A) to remove comment about _col_
B) to re-add the example you removed
... I would prefer B), but you already removed it so, you probably prefer something else :-)